bam_consensus.c

```diff
@@ -536,16 +1032,36 @@ int nm_init(void *client_data, samFile *fp, sam_hdr_t *h, pileup_t *p) {
             }
         }
         for (; i < qlen; i++) {
-            int t = (qual[i]   + 5*qmin)/4; // good on 15x
-            local_nm[i] += t < qual[i] ? (qual[i]-t) : 0;
+            int t = (opts->mode == MODE_BAYES_116)
+                ? (qual[i]   + 5*qmin)/4
+                : qual[i]/3 + qminp*poly_adj;
+            local_nm[i] += t < qual[i] ? qual[i]-t : 0;
         }
-#endif
     }
 
     // Fix e.g. PacBio homopolymer qualities
     if (opts->homopoly_fix)
         homopoly_qual_fix((bam1_t *)b);
 
+    // local_nm[i] & ((1<<24)-1) is for SNP score adjustment.
+    // We also put some more basic poly-X len in local_nm[i] >> 24.
+    uint8_t *seq = bam_get_seq(b);
+    for (i = 0; i < qlen; i++) {
+        int base = bam_seqi(seq, i);
+        int poly = 0, j, k;
+        for (j = i+1; j < qlen; j++)
+            if (bam_seqi(seq, j) != base)
+                break;
+        //printf("%d x %d\n", base, j-i);
+
+        poly = j-i-1; if (poly > 100) poly = 100;
+        const int HALO=0;
+        for (k = i-HALO; k < j+HALO; k++)
+            if (k >= 0 && k < qlen)
+                local_nm[k] |= (MAX(poly,local_nm[k]>>24))<<24;
```

Comment:
This works for `HALO=0`, but the update would need to be done rather more carefully if it was bigger...

Reply:
I could cull that, given it was a failed experiment that didn't help much, but I don't understand what's wrong with it.

Basically it changes a loop that goes from `i` to `j` (coords of the homopolymer) to `i-halo` to `j+halo` to include a few bases either side of the homopolymer.  This means we adjust `local_nm` not only for "within a homopolymer" but also "adjacent to a homopolymer".  It sounded like a reasonable thing to consider, but didn't really pan out.  I left it there incase different aligners and/or instruments demonstrated there are reasons to do this, in which case it'd become a parameter.

There is bounds checking in the code though, regardless of `HALO` value.

It gets set into the top bits of `p->cd` so the `poly_len` function can skew base quality.  The amount of skew is the `--homopoly-redux` option, which potentially has some benefit to ONT.  However it's quite a tradeoff of FP vs FN and for simplicity I decided to leave it set to a constant amount and left it out of the documentation and usage statements.  Basically it may help, but the jury is out and I didn't want to lose the code.

So for now it's just a fixed non-zero factor (0.01), which essentially just means decrementing qual by 1 if poly-len is > 2.
I could ifdef out all code involved if you prefer.

Reply:
You're using bitwise-or to set the new value.  That's OK if the old one was zero, but if you allow `k` to extend outside the range `(i .. j]` then the top bits of `local_nm[k]` might have been set by an earlier pass through the loop.  If that happens, the bitwise-or may set the wrong value depending on exactly which bits are set.